### PR TITLE
[react]: Infer state type for useReducer

### PR DESCRIPTION
**Pull request by mfp22**
- [ x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ x] Test the change in your own code. (Compile and run.)
- [ x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [ x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [ x] Provide a URL to documentation or source code which provides context for the suggested changes: [<<url here>>](https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/63607)

### CODE REVIEW

File: types/react/index.d.ts
Comment (eps1lon) [inline types/react/index.d.ts:989]: We can remove NOTE and NOTE 2 now. 

I can't even get autocomplete to work with the previous typings (down to TS 3.3): [Playground with previous types](https://www.typescriptlang.org/play?ts=3.9.7#code/JYWwDg9gTgLgBAbwK4GcCmAlNATJBjNKAXzgDMoIQ4ByKNAQzxmoCgWB6duGNFGF4ADseUUozRwAgmDABlGPR6IWcOIPog0ALjh8oQgOYBuFXHoHtapCABGhE0TYwAnmAnSwkpsAiCUcAF5TAB9EbldLACILGAB5ABtsQki4R1VQhHC3HUi6dBhJCxSiExZSJEFvXzg6XAIoAAoUHQ95RTQAGjMqwRaZLxgfPwBKPrkFJQRTFAB3YBg8AAs4BsZB3wA6Fzdh5VVVPHp0OGi0OMTkrVN9mrOkKEEwjeeULvNLFA33uABqOABGVImG6HY65XhnQpoSJXG6qOgwe6PTLPDavMwWHQABiBpkcjjYqEwOHwhAatVJUC6mXeWgALAAmLqpYZGIA)

Proposed typings don't have autocomplete either: [Playground with proposed types](https://www.typescriptlang.org/play?#code/PTAECcFMEMGMBcB0ATR8DOAoTJTwBYCW6eks+AdobNADa0CeoyA9pCXLJAA7yjSh0ZFhWT9wAcwCuAW0gV4AGlAAjKX0LwA5B1pRoyJlNGRw-Zjyg14hEaADu0cFQoScYaKNCadoCiz5IADd5CEhaGCExdBZQbnAWFWgVRlVIeHhTPFj-PmRIADNCCkhvJEx4Bm5SgBFibmh4cgAeAEEAPlAAXlAACiC6KUgALlBWgEpuzqCWQmQAbndQAGVirn4EWwpQAH0aCh200GN8opLkZWR6xvJQGWgmFVKaekgxe018FnV+CiYG8DQOSZcDoRAVKq1a5NfAAdU+33grU2dh6vUmXWmswWSwAqhRaIQANalKDIKQAD2UBFKcBsIhIe08hyevwYBGKbkq1VAACU3lIuOBmstlB1un14sFlvBGiMVso6VtRhMpitFrgAJIFY5CfnkoUbLi8DhhA1ZD4EREbekUS7Q273R7POgRd4In6ef5OIHpUxgiE8-WC0zwq3qZG2kWdNFSoIyuWjZYYzrLDVgbnsXVvbJ4cBMTxiOPyPg00CwFgybiECJmAoJGRmwWclZLeCxAToKTVcCZ8uVmR2S34PD4UpCCteJzSOQKXNSPUCoXowOlYNChOZZq80CQCmZUQkdemZpexV-doxvm7-fyZBHpcn4oFLKitmdAD8K1AoxKIXAix9se4CRls243ge958o+wpnu+V47nukEPuasF-Moz5ZOKX6tD+fjBKY6agAAKmO3j5AomhMOQZBEt4JCsCUDifKOpTLL6ABqgy0rQEgsOAnyNr0ADyKgAFZkEgxDjMo9hELcxCCPAAkIFkBCeN0XRdKu0GoZukBhl8EYohQ4FIXeKEhuAhmIqBIinheMaYKA17mYeulWTZxlRphZjLJ+36-gRAHYMAABUYXOS5LlhWM2x0CCFCNIQIS5gABgukD6Wl4LRTFUXRbFGWLqhaX0bqUiuv8UAvoCKSlO2oDFVlsqZGVcmhAw3ygPg0CpRWVYRBSSlyqAtAsBI1Cjo03gUDMtAhFgeWgLFMhSLQNjcBEghSCoAC0Ay0EMYKgJqfB0DEY3pCQXVSKALC8IQMiEAAXqUPYFPx9wUOsn1mAN3AiCWJDqaWAkSBIWT5DwBUxcc3DIHKJBPDQmWgLd5YaQ06AkGlVzoA0MJlaw9jbMU6CZAY906i8tBJLARIBstkXM6AAAC-7oFsoAAIwAGyIAAHIgAAMsMrezQilPgGTcOgwwgPoCBiWC-ESMArCwOgwBfCwjN7TVpjyFwiAyzItAAMSZWSVni2FwBLCw-7jVTHVQMwDojhW61iJwPB8CL4gzsD4L5LAtBOKUBTGCZ2bAWZt7ucBXlIiZDkMO0yiau0vRRTbQqjLyihRcUmiEHQr2mK0kijJqxcuaXNgV294CjL0061ymHkbq1BkeqnUa8u0mDjKMADawH6SndmmUPyh1ATNxwv3M8ALoahFzmxa08UbaYSU2KljXNdluUs0VmXAWVikLpV9DVYUpjJNtx+ZdlDhjtsGO9f1lZbXuI1MhjQmlNUGs15qLS3ncdam1tpdn2odY6iBTrnVoJdCIGB0bdQejYZ6b04imD+t9X6-F+xViBgoEGvUwaEAhlDSAMMJbdkRpkZGZBoBowxvsOI0AcZNXxoTcgxMWCk1mhTGAYgWA01dPTRmZ8oEc39NzfmQtRYKKlj1WW8tFYwGVqrSQGsWBax1iwPW6ADaPygD9SApt4DmytkIfOpgt4O1wE7UwLt3hjndgIpe-YfZGn9qAQO05ZAh0wGHCO7to4-VtHHGCCdkLd1DCvNOXpLy5xck41ufJ66zTLs3KuNdkngCnqkwemcS5UCboSFuH5RgnEKMUN4I9x6T17tPNOc9QAL0EcvcMA8tjrxwJvCWO9+B72cMlI+sQT69xylvKBzUr7lVvlVOIhs6ov1mW-eZH9OrdR-s8P+Q1AGlHGpNWA00NBzRYAtdgUC1obUIP-HaCDuInTOhM9B10sF3RwU9SuBDwBEM8CQ-6f8KGYLAcpWhkMzDQ24FA5hSM0ioyEH8zG2xsa418UTZgIiyYUHEVTKRmN6CyKZufdmnNlEC2FmLCWbMNEy3gHLBWwAlbwBVogNWhjjG631ps42NizaW2tjBFxjtnYsFdt40oAAiTUCq7gPCOJ2XaQg+BkvaXKbc7R5jZjEI1eITs5i0nUEYk56Qtjgi1DqJVKr7gwizLqrcQ8bx0lSDSbYZYI4U3ujK12NZaATPGvYfJNTK5mEaqySsmhMioDbGRf12qg1iEKC+TYIRUhPCORwEghAdRlnNZRSooAYmx16MNTEoAKSTEUiW6p5dammFDmQKJUcY5xMvgkxCicoLxzgukrOOc84wULnkxuLbo3VwkLXUAAAyEp+l9VTubYUnJ7dimaiXSu3u+qu5usgPq1poAJ4wVXd03pS9twwRnvq9owzwrUvGQlfe0yGo7KEKfRZEtlkwWviQNZ98NmWOfl+pquy5TtU-pio5ZD-7DQpqNC5oDqHgLuZAiWTzYHjl2gdD5yCvkXViBgm62DHp4PeoQr6YKo6kIBlCqhM1YV0IRQwpFTCEaopRhwjFXCsa8NxZ7YRojyaU0kdIilcA5F-tioo0EdLVGMoUyyrRHKuU8r5ZrbWgrzHCusbY+xErUJSrcemitpD6wMNAI6w1rpQB8SzLw4Brh+CFo3a2iFc1TCYM0LNRqCrj36oVe28OkcK3du5r21CiSLIlPTueDOo7MlNgLrkqpBTvNztruu7L0a24d1Okey9B6h5novXpcrmcemezvahB9Q8n0b1fbvRKn70rQbavI-9sWrJAYqusqUtUINdZ-XsjqX9Dl9WOYNABKGgFoauWA4oECHk4ZgS8uBBHEHsGI6gn5mCMYAuo8C0F1irMQvISUSh1y8xwvoYw2KKLWFov46UQT2LhP8NEwS8TxLJPU3JXTWTVKFG0rsCohl6jbOsvZTouk2mDG6ZMWYixtURXGfFY4yVK0HZLAAHLCWIgAUVGMOa0ZZj2zSuFYW0yhiKQmWLAASvAHDfFoGIJx35Y0NTIoOANA1BwUCWF2HsfYnjwHsAw31ZFsm+CgPAKQzg8CQl+Masi066D6T6KQptBWW6K-SCr31kIZJLDktQEc9hOdFigCEOcHD2wAwwdzeslZdz-g5+AIknJwRE5J6T0AAAmUYpF3q8L4JOIU8UEgnH4Ja13NqhwsTLEr03OZFulESBJBAYuGEyBBjkMjIh4XqrJjjIYyDlhw40yASaBBdqIAGsAGoTTqnhAYEz6oyA28d80F3nvbwNbEFgAuLmDJgB8wAMx8xFgAdgiR2yLlae0lSsiKMUY6skTr6PLBU-AVRd1FFlmp+kkyVbfO3Eyx+1QzDmMM0KGZ2DwEwMUEEBQ4ClFaNwbguuABvKKJKOQUYCmASVwRYFyaASGX8WQJ4EKAAX2wD7F-24BnhIG0hcgAB9QAAC1dqhRgFVIZ4BhIudTAVVkCcC8CCD5QFUoAtVWhIZKDFhMA19uZslegD80D9JFRb8xg-8MDR4BD-9e48Cop0BLRbgb9bQ0BzdxC8oaAMViD0gyD8hwAFVhhxYwhldVd8DEADD0BFRYDBBEAYDSgABqXmUARAqA6KJQxVBg9IJgyATQ7QjPPQ0AAwxAIw-gEwwOWwqKZA5A7AfrZcbJZQfA8w0YAAFhD2UEQPGHmCAA)

Comment (mfp22) [reply types/react/index.d.ts:989]: Strange. It works in the StackBlitz I made. If you go to line 40 and erase the `count` property and start typing again it autocompletes it. https://stackblitz.com/edit/react-ts-gy5gv6?file=App.tsx

Anyway, pipeline failed so I'm exploring that. And I'll erase those notes

Comment (mfp22) [reply types/react/index.d.ts:989]: My overload is being used but the other specific overloads are causing my overload to not be able to infer the type. I will explore this more.

Comment (mfp22) [reply types/react/index.d.ts]: @eps1lon  I came up with something that might work. 2 overloads instead of 6. I just updated this PR.

Tests pass locally _except_ for the test I added in this PR. I don't know why, because in VSCode running 5.0.0-dev20221218 It's correctly inferring how I want so the test should pass, and in the TS environment like what you shared, it's inferring correctly with v4.9.5. If it's failing for me locally, I bet it will fail in the pipeline. If it does fail, should I just remove my added test? It will pass existing tests and apparently _sometimes_ be better than what's there now. 

In VSCode it's autocompleting, and I just confirmed it is in [the TS environment](https://www.typescriptlang.org/play?#code/PTAECcFMEMGMBcB0ATR8DOAoTJTwBYCW6eks+AdobNADa0CeoyA9pCXLJAA7yjSh0ZFhWT9wAcwCuAW0gV4AGlAAjKX0LwA5B1pRoyJlNGRw-Zjyg14hEaADu0cFQoScYaKNCadoCiz5IADd5CEhaGCExdBZQbnAWFWgVRlVIeHhTPFj-PmRIADNCCkhvJEx4Bm5SgBFibmh4cgAeAEEAPlAAXlAACiC6KUgALlBWgEpuzqCWQmQAbndQAGVirn4EWwpQAH0aCh200GN8opLkZWR6xvJQGWgmFVKaekgxe018FnV+CiYG8DQOSZcDoRAVKq1a5NfAAdU+33grU2dh6vUmXWmswWSwAqhRaIQANalKDIKQAD2UBFKcBsIhIe08hyevwYBGKbkq1TGKIodXQDRhzQASm8pFxwGL0IEKZlRBw-gBtAC6nS6mFAWtAYvJkulsvlyBISs8DBVoAA-KABUKWrqJaYDUqAAxq0CjW03OEI9TI+kURbYYAAKhDmpDY22dBBFEahBC2VAAAMpEJlvBGpBk+DQOG8ym05AHZLk94SGmpHRUvFCqZkhEk6n05nMmX7PhQgxvqB8NBE7AWDJuBEKYJW6VaCwJNQ8H2NBQZrQQlgCzIpLQbCPSugpCoALQDWhDMGgACSfDoMVAEQwoG7UlALF4hBkhAAXqVquACixwPcKHWX8zEHYcRHkO8CEaPBwEICQJCyfIeAjY5uGQLMSCeGgi3vHt9jiaB0BIZMrkFb0y1YextmKGUYDEFgClAF5aCSWAiTBCMUIAARCUEtlAABGAA2RAAA5EBdbihFKfAMm4dBhhAfQEAAKzBP8JGAVhYHQYAvhYdj9ygApTHkLhEFkmRaAAYiLMlHXACNgEwfJYFoJxSgKYw+WOIQS1MZplmUVpQEgOV5GNNlVXaXpNW1ezJVGXpayCDMs1GILQEQbKnAkBSxgxTogrirVik0Qg6DSzIMswcZRiVTL-S2L1hQ6UAVUWUN80jVpo03Uw4xsRN4FiZtICq7Nc26ws-PFUty18qt6H+Yz6xSUoRpm8aJ3bTttgfXt+2eIdtzHGUsxvadZyghclxXFD103Qht0EPdD0GdhEHPS9aGvW8SAO58bDfT84lMYCAKAv8mJO8CFBIG6YLghCzCQ7gUKkNCMLSbChFwx98IaIiU1Iu18AolgqO8ChaIMJ9GOY1j2Km7jePQfjhLEiSpMgGS5IUpSYFU9TJC0lgdL0lgDPQIy6ygQDIAs+ArNsoQEtMJyXLIdyoFALzAIDXzizmgLMrPYLQvChUorVWLtTCPVTCSlKJoy5RssQXL8omKYVkUErqfKuhWkkUZzYDsr4CSwgw8KlZavqxq+RaloOg67AlkyGVMGKEECjgUpWm4bgJtAABvOK4zkUYZVg1xFi1aAENGChZCecBFgAX2wblC+LpqGW6OKAB9y7wSFRgAIgQ+AAHlaHycBJ9AbutVHsvx+qKeoCEJEEOXzug0wfWfPV8Bem94uJuUOktlGIvuAHmm6rGK+J3LuL0A+GE+lvkQ0EhJMCu9saB42nukeei9J7DADlqKA8ApDODHh7dAN9m6CE9ghUAABqQSK8G7alAaUSeO90itH3jA+2cD0iIO2BvFBaCRigBdPguK3du5LAAAoJCCHMUokcKq0HHFmZQxQTJmDPsIzIm9ICYCLP5c+cUT6GzPhfF+D9S7AKofA2hyDsqoP4OgsETdSi4IEqwrUnd-Zag3iY0YAAWAATMoKuTCtA1E8IQcIWhu7jCDIOGmfAlTnUyBaHo8iTaKK1BfG+fJ779z5OgdRb8LqYg-vbL+mhbi9D-hQAB1QgGwKYoRYhM9IGmGgUU6hCCkH0P0YwmumDTF4MPkUohoASHsDIRQqpYQal0KyvUwxTCWGtPtqvFe1ix52NAE4lxQI3EeKoN4letUgwhMViY4MYAAAqkIHYOSkRtSEN91DixOrefhVAbB0COXI2ajsol9EviXCcyTXmpM6Fo+2OjamDMQAYmZxisFmIsZMuKtj0FzL8As0Y7jPHeN8f4hkQSNmOLCUbBRdsDmSimZCph0LXFwqWV42gWhVl+OwBsppQA) too.

Comment (eps1lon) [reply types/react/index.d.ts]: Less overloads is always a win. Especially if we can have 3 or less overloads since that has poor debuggability: https://github.com/microsoft/TypeScript/issues/52745

Comment (mfp22) [reply types/react/index.d.ts]: Yes it failed in the pipeline for some reason. I'll try to figure it out a bit longer, but I could just remove my test and make this a "simplify useReducer overloads" PR (maybe it fixes autocomplete too) and in every environment except whatever this specific one is, it will infer the reducer state like I wanted. 
image

Comment (mfp22) [reply types/react/index.d.ts]: I got it passing everything locally! `$ExpectType` is very picky and not helpful at telling you how you're using it wrong. But now I know.

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## react ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react@latest/))
was missing the following properties:
1. unstable_act
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 0de4eea111d14277e5d004ac9768ad18e9399450
</p>

**Review by eps1lon — CHANGES_REQUESTED** [review]
Super excited to see this land while also reducing the number of overloads. 

I need to double check, but I think this might cause some breakages if people specify the type parameters since some signatures are now gone. But I have to double check. Want to test this on some larger repos anyway before landing.

Leaving the requested changes mostly, so this doesn't get merged before we do some integration testing.

File: types/react/index.d.ts
Comment (mfp22) [reply types/react/index.d.ts:865]: Yeah I'll make all of this into a single commit with only necessary changes and force push it up. I'll get this done later today.

Comment (eps1lon) [reply types/react/index.d.ts:865]: No need to squash in a PR. Oftentimes better to keep single commits to understand the work progression. We squash on merge anyway.

Comment (mfp22) [reply types/react/index.d.ts:865]: Well I'm not super great at Git... I was reading up on how to do this and I ended up squashing just the last 3 commits and amending the remaining commit with changes back to how things were with these types/comments. I also restored the `ReducerWithoutAction` type because it actually is a useful type on its own, and that would probably be where breaking changes could occur. And now VSCode is showing the original authors on the lines I originally deleted. So I changed the commit history on my local branch, so my bad with that. I force pushed it just a minute ago. 

This is actually my first PR to an open source project. I have my own state management library (as one does) called [StateAdapt](https://state-adapt.github.io/react) but I've just been committing directly to `main` until other start contributing.

Comment (mfp22) [reply types/react/index.d.ts:860]: I just realized this comment no longer totally reflects why this type exists. It's not for `useReducer` anymore. Should I update this?
```
    // If useReducer accepts a reducer without action, dispatch may be called without any parameters.
    type ReducerWithoutAction<S> = (prevState: S) => S;
```

Oh and I just remembered I've opened PRs to random small projects on GitHub. Just nothing serious that more than 100 people have starred.

Comment (mfp22) [reply types/react/index.d.ts:860]: Actually that comment still applies if they use it to type a reducer defined outside of `useReducer` that's still used with `useReducer`... idk

Comment (eps1lon) [reply types/react/index.d.ts:860]: Yeah this was more about (needlessly) moving around types. Now we have a clear diff showing removed and added types.

**Review by eps1lon — CHANGES_REQUESTED** [review]
I love this change and it integrates pretty well (basically just removing explicit type parameters which is nice). However, the changes to the type parameter interface will require work. 

That's why I think we should schedule this for the next React major versions so that we don't break any code relying on the current type parameters. I added it to https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/64451 so that we don't forget.

One thing I noticed when integrating is the following issue:

```ts
const reducer = (state: boolean, next?: unknown) => {
  if (typeof next === 'boolean') {
    return next
  }

  return !state
}

const useToggle = (initialState: boolean) => useReducer(reducer, initialState)

const [isTooltipVisible, toggleTooltipVisibility] = useToggle(false)
toggleTooltipVisibility(false)
//                      ^^^^^ Expected 0 arguments, but got 1
```

Is this something we can fix?

**Comment by mfp22** [comment]
> That's why I think we should schedule this for the next React major versions

Oh that could be years away... too bad. Sure we can't just YOLO this into React 18? I'll volunteer to fix everyone's problems :)

> ```ts
> const reducer = (state: boolean, next?: unknown) => {
>   if (typeof next === 'boolean') {
>     return next
>   }
> 
>   return !state
> }
> 
> const useToggle = (initialState: boolean) => useReducer(reducer, initialState)
> 
> const [isTooltipVisible, toggleTooltipVisibility] = useToggle(false)
> toggleTooltipVisibility(false)
> ```

The only way I've found to solve this case is to spread the action arguments onto the dispatch function directly. That preserves optional parameters, but would allow potentially multiple arguments to be passed into `Dispatch`. 

So, I changed `useReducer` to limit the number of reducer parameters. This matches the Flow types better as well as the prior behavior here.

I added tests to reject reducers with extra parameters, and a test to make sure optional action parameters are preserved.

If someday they change `useReducer` to take in a 3rd argument for some reason (in [StateAdapt](https://state-adapt.github.io/docs/core#adapter) I pass in `initialState` to the reducer in the 3rd argument, which is really useful, so it could happen) then we can change the name `AnyActionArg` back to `ReducerRest` or something and allow it to be length 2 or whatever as well. There are ways to handle it. We can always check against `ReducerRest['length']` to narrow down to specific types.

**Comment by weswigham** [comment]
Draft should be good - that aughta remove it from the maintainer queue.

**Comment by mfp22** [comment]
@eps1lon I don't understand how this build could have failed with just the changes in this PR. It looks like it doesn't have anything to do with `useReducer`. https://github.com/DefinitelyTyped/DefinitelyTyped/actions/runs/4776139211/jobs/8490968425?pr=64412

I also noticed a `ts5` folder. Should I add my changes to that folder too?

**Comment by mfp22** [comment]
@eps1lon Does this mean this PR can be resurrected soon? https://www.reddit.com/r/reactjs/comments/1ajur2f/is_there_any_news_on_react_19/

**Comment by eps1lon** [comment]
It will be part of the React 19 release. I already have a branch that includes the commits from this PR.

**Pull request closed without merging**
